smartconfig/entry.py

```diff
@@ -0,0 +1,153 @@
+        for attribute in chain(cls.__dict__.keys(), getattr(cls, "__annotations__", ())):
+            if not attribute.startswith('_'):
+                try:
+                    get_attribute(cls.__path, attribute)
+                except (ConfigurationError, ConfigurationKeyError):
+                    raise ConfigurationKeyError(f"Attribute {attribute!r} doesn't have a defined value.") from None
+
+    def __init__(
+            cls,
+            name: str,
+            bases: Tuple[type, ...],
+            dict_: Dict[str, Any],
+            path: Optional[str] = None,
+            defer_attribute_check: bool = False
+    ):
+        """Initialise the new entry."""
+        super().__init__(name, bases, dict_)
+
+        cls.__path = path or cls.__module__
+        if cls.__path in _registry.used_paths:
+            raise PathConflict(f"An entry at {cls.__path!r} already exists.")
+
+        cls._register_entry()
+        if not defer_attribute_check:
+            cls._check_undefined_entries()
+        else:
+            _unchecked_entries.append(cls)
+
+    def __del__(cls) -> None:
+        """Remove the default values from the global configuration and free the entry's path."""
+        try:
+            unload_defaults(cls.__path)
+        except (ConfigurationError, ConfigurationKeyError):
+            # Fail silently because the same error may have already been raised while the object was being initialised.
+            # There's no way to distinguish such case from the state being modified after successful initialisation.
+            # Besides, there's nothing that can be done for cleanup if the node can't be retrieved.
+            return
+
+        # If the node has already been cleaned up we don't want this to error out
+        with contextlib.suppress(KeyError):
+            used_paths.remove(cls.__path)
+
+    def __repr__(cls) -> str:
+        """Return a short representation of the entry."""
+        if cls is ConfigEntry:
+            return super().__repr__()
+        else:
+            return f"<ConfigEntry {cls.__qualname__!r} at {cls.__path!r}>"
+
+
+class ConfigEntry(metaclass=_ConfigEntryMeta):
+    """
+    Base class for configuration entries.
+
+    Values of class attributes can be overwritten by loading YAML configuration files. Otherwise, their given default
+    values will be used. The exception is class attributes whose names begin with an underscore; they always behave like
+    normal class attributes and therefore their values cannot be overwritten by config files.
+
+    Metaclass Args:
+        path: Path to use for attribute lookup in the config.
+            Default to the containing module's fully-qualified name when the value is None or an empty string.
+        defer_attribute_check: Whenever checking for missing attributes should be done at a later date.
+            Use `check_attributes()` to perform the check. Default to False.
```

Comment:
Similar to the suggestion for the load functions.

```suggestion
        check_attributes: True if a check should be performed for attributes without defined values.
            If False, `check_attributes()` can be manually called later. Default to True.
```